allennlp/data/dataset_readers/encoder_decoder.py

```diff
@@ -0,0 +1,105 @@
+        Indexers used to define input (source side) token representations.
+    target_token_indexers : ``Dict[str, TokenIndexer]``, optional (default = self.source_token_indexers)
+        Indexers used to define output (target side) token representations.
+    """
+    def __init__(self,
+                 source_tokenizer: Tokenizer = None,
+                 target_tokenizer: Tokenizer = None,
+                 source_token_indexers: Dict[str, TokenIndexer] = None,
+                 target_token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._source_tokenizer = source_tokenizer or WordTokenizer()
+        self._target_tokenizer = target_tokenizer or self._source_tokenizer
+        self._source_token_indexers = source_token_indexers or {"tokens": SingleIdTokenIndexer()}
+        self._target_token_indexers = target_token_indexers or self._source_token_indexers
+        self._token_indexers = {"source": self._source_token_indexers,
+                                "target": self._target_token_indexers}
+
+    @overrides
+    def read(self, file_path):
+        instances = []
+        with open(file_path, "r") as data_file:
+            logger.info("Reading instances from lines in file at: %s", file_path)
+            for line_num, line in enumerate(tqdm.tqdm(data_file)):
+                line = line.strip("\n")
+
+                if not line:
+                    continue
+
+                line_parts = line.split('\t')
+                if len(line_parts) != 2:
+                    raise ConfigurationError("Invalid line format: %s (line number %d)" % (line, line_num + 1))
+                source_sequence, target_sequence = line_parts
+                instances.append(self.text_to_instance(source_sequence, target_sequence))
+        if not instances:
+            raise ConfigurationError("No instances read!")
+        return Dataset(instances)
+
+    @overrides
+    def text_to_instance(self, source_string: str, target_string: str = None) -> Instance:  # type: ignore
+        # pylint: disable=arguments-differ
+        tokenized_source = self._source_tokenizer.tokenize(source_string)
+        source_field = TextField(tokenized_source, self._source_token_indexers)
+        if target_string is not None:
+            tokenized_target = self._target_tokenizer.tokenize(target_string)
+            tokenized_target.insert(0, Token(START_SYMBOL))
+            tokenized_target.append(Token(END_SYMBOL))
+            target_field = TextField(tokenized_target, self._target_token_indexers)
+            return Instance({"source_tokens": source_field, "target_tokens": target_field})
+        else:
+            return Instance({'source_tokens': source_field})
+
+    @classmethod
+    def from_params(cls, params: Params) -> 'EncoderDecoderDatasetReader':
+        source_tokenizer_type = params.pop('source_tokenizer', None)
+        source_tokenizer = None if source_tokenizer_type is None else Tokenizer.from_params(source_tokenizer_type)
```

Comment:
I see that there are two kinds of defaults. When a choice is not specified in the json file, the classes processing those params have two options: either make a default choice themselves, or having the class relevant to the choice (like `Tokenizer`) make a default choice. I think what the `Tokenizer` is doing is fine. But it would be nice to have a cleaner way of doing the former as well.